src/main/java/build/buildfarm/instance/AbstractServerInstance.java

```diff
@@ -434,7 +438,11 @@ protected void updateOperationWatchers(Operation operation) {
 
   @Override
   public Operation getOperation(String name) {
-    return outstandingOperations.get(name);
+    Operation operation = completedOperations.get(name);
+    if (operation == null) {
```

Comment:
You have it exactly right, both in intent and the likelihood of a race - this is a textbook transaction and needs to be treated as such - I'll update further in a bit.